Explorer/Assets/DCL/SDKComponents/MediaStream/Systems/VideoEventsSystem.cs

```diff
@@ -0,0 +1,81 @@
+﻿using Arch.Core;
+using Arch.System;
+using Arch.SystemGroups;
+using Arch.SystemGroups.Throttling;
+using CRDT;
+using CrdtEcsBridge.ECSToCRDTWriter;
+using DCL.Diagnostics;
+using DCL.ECSComponents;
+using DCL.Optimization.PerformanceBudgeting;
+using DCL.Optimization.Pools;
+using ECS.Abstract;
+using ECS.Groups;
+using RenderHeads.Media.AVProVideo;
+using SceneRunner.Scene;
+using System;
+
+namespace DCL.SDKComponents.MediaStream
+{
+    [UpdateInGroup(typeof(SyncedPostRenderingSystemGroup))]
+    [LogCategory(ReportCategory.MEDIA_STREAM)]
+    [ThrottlingEnabled]
+    public partial class VideoEventsSystem : BaseUnityLoopSystem
+    {
+        private readonly IECSToCRDTWriter ecsToCRDTWriter;
+        private readonly ISceneStateProvider sceneStateProvider;
+        private readonly IComponentPool<PBVideoEvent> componentPool;
+        private readonly IPerformanceBudget frameTimeBudget;
+
+        private VideoEventsSystem(World world, IECSToCRDTWriter ecsToCrdtWriter, ISceneStateProvider sceneStateProvider, IComponentPool<PBVideoEvent> componentPool, IPerformanceBudget frameTimeBudget) : base(world)
+        {
+            ecsToCRDTWriter = ecsToCrdtWriter;
+            this.sceneStateProvider = sceneStateProvider;
+            this.componentPool = componentPool;
+            this.frameTimeBudget = frameTimeBudget;
+        }
+
+        protected override void Update(float t)
+        {
+            PropagateVideoEventsQuery(World);
+        }
+
+        [Query]
+        [All(typeof(PBVideoPlayer))]
+        private void PropagateVideoEvents(ref CRDTEntity sdkEntity, ref MediaPlayerComponent mediaPlayer)
+        {
+            if (!frameTimeBudget.TrySpendBudget()) return;
+
+            VideoState newState = GetCurrentVideoState(mediaPlayer.MediaPlayer.Control);
+
+            if (mediaPlayer.State == newState) return;
+            mediaPlayer.State = newState;
+
+            ecsToCRDTWriter.AppendMessage<PBVideoEvent, (MediaPlayerComponent mediaPlayer, uint timestamp)>
+                (PrepareMessage(), sdkEntity, (int)sceneStateProvider.TickNumber, (mediaPlayer, sceneStateProvider.TickNumber));
```

Comment:
This way you allocate the delegate everytime you call the method - it's wrong - you must ensure it's cached: either automatically with a static lambda (neither static method nor a method that returns a delegate as in your case) or manually by caching a delegate in the constructor